module/zfs/bpobj.c

```diff
@@ -242,119 +277,171 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+	list_create(&stack, sizeof (bpobj_info_t),
+	    offsetof(bpobj_info_t, bpi_node));
+	for (bpobj_info_t *bpi = bpi_alloc(bpo, NULL, 0, 0);
+	    bpi != NULL; bpi = list_remove_head(&stack)) {
+		bpo = bpi->bpi_bpo;
+		mutex_enter(&bpo->bpo_lock);
+		ASSERT(bpobj_is_open(bpo));
+		if (free)
+			dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+
+		err = bpobj_iterate_blkptrs(bpi, func, arg, tx, free);
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
 			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			goto out;
 		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
+		/* A subobj with no subobjs below it can be cleaned up. */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
+			}
+			if (free && (bpi->bpi_parent != NULL)) {
+				bpobj_t *p = bpi->bpi_parent->bpi_bpo;
+
+				ASSERT0(bpo->bpo_phys->bpo_num_blkptrs);
+				ASSERT3U(SUBOBJS(p), >, 0);
+				ASSERT3U(bpi->bpi_index, ==, SUBOBJS(p) - 1);
+
+				p->bpo_phys->bpo_num_subobjs--;
+
+				VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
+				    p->bpo_phys->bpo_subobjs,
+				    bpi->bpi_index * sizeof (uint64_t),
+				    sizeof (uint64_t), tx));
+
+				/* eliminate the empty subobj */
+				if (bpo->bpo_havesubobj &&
+				    bpo->bpo_phys->bpo_subobjs != 0) {
+					ASSERT3U(SUBOBJS(bpo), ==, 0);
+					err = dmu_object_free(bpo->bpo_os,
+					    bpo->bpo_phys->bpo_subobjs, tx);
+					if (err)
+						break;
+					dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+					bpo->bpo_phys->bpo_subobjs = 0;
+				}
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
+				if (err)
+					break;
+			}
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			continue;
 		}
+		ASSERT(bpo->bpo_havecomp);
+		/*
+		 * Can't clean up bpobj while it still has subobjs.  Place it
+		 * back in the list ahead of its children.  It will be cleaned
+		 * up once the children are all freed.
+		 */
+		if (bpo->bpo_phys->bpo_num_subobjs > 0)
```

Comment:
I think num_subobj has to be nonzero here (otherwise we'd have cleaned up the bpo and `continue`d above).  We wouldn't want `i` to be `-1` below either.